golang/cosmos/x/swingset/keeper/snapshotter.go

```diff
@@ -0,0 +1,421 @@
+	}
+
+	if snapshotter.app.SnapshotManager() == nil {
+		return fmt.Errorf("snapshot manager not configured")
+	}
+
+	// Set the structure used to synchronize with the goroutine
+	// its presence indicates a snapshot has started.
+	snapshotter.activeSnapshot = &activeSnapshot{
+		height,
+		make(chan error, 1),
+		false,
+	}
+
+	go func(startedResult chan error) {
+		action := &snapshotAction{
+			Type:        "COSMOS_SNAPSHOT",
+			BlockHeight: snapshotter.activeSnapshot.height,
+			Request:     "initiate",
+		}
+
+		_, err := snapshotter.blockingSend(action)
+
+		if err != nil {
+			// First indicate a snapshot is no longer in progress if the call to
+			// `WaitUntilSnapshotStarted` has't happened yet.
+			snapshotter.activeSnapshot = nil
+			// Then signal the current snapshot operation if a call to
+			// `WaitUntilSnapshotStarted` was already waiting.
+			startedResult <- err
+			snapshotter.logger.Error("failed to initiate swingset snapshot", "err", err)
+			return
+		}
+
+		// Signal that the operation has started in the goroutine. Calls to
+		// `WaitUntilSnapshotStarted` will no longer block.
+		startedResult <- nil
+
+		snapshotter.app.Snapshot(height)
+
+		if snapshotter.activeSnapshot.retrieved {
+			snapshotter.activeSnapshot = nil
+			return
+		}
+
+		snapshotter.logger.Error("failed to make swingset snapshot")
+		action = &snapshotAction{
+			Type:        "COSMOS_SNAPSHOT",
+			BlockHeight: snapshotter.activeSnapshot.height,
+			Request:     "discard",
+		}
+		_, err = snapshotter.blockingSend(action)
+
+		if err != nil {
+			snapshotter.logger.Error("failed to discard of swingset snapshot", "err", err)
+		}
+
+		snapshotter.activeSnapshot = nil
+	}(snapshotter.activeSnapshot.startedResult)
+
+	return nil
+}
+
+// Synchronize with a snapshot in progress, if any.
+// Waits for a just initiated snapshot to have started in its goroutine.
+// If no snapshot is in progress (`InitiateSnapshot` hasn't been called or
+// already completed), or if we previously checked if the snapshot has started,
+// returns immediately
+func (snapshotter *SwingsetSnapshotter) WaitUntilSnapshotStarted() error {
+	// First, copy the synchronization structure in case the snapshot operation
+	// completes while we check its status
+	activeSnapshot := snapshotter.activeSnapshot
+
+	if activeSnapshot == nil || activeSnapshot.startedResult == nil {
+		return nil
+	}
+
+	// Indicate to future callers that we already waited for the snapshot
+	// operation to have started so they can return immediately.
+	startedResult := activeSnapshot.startedResult
+	activeSnapshot.startedResult = nil
+
+	return <-startedResult
+}
+
+// SnapshotName returns the name of snapshotter, it should be unique in the manager.
+func (snapshotter *SwingsetSnapshotter) SnapshotName() string {
+	return types.ModuleName
+}
+
+// SnapshotFormat returns the default format the extension snapshotter use to encode the
```

Comment:
```suggestion
// SnapshotFormat returns the default format the extension snapshotter uses to encode the
```

I also find the `SnapshotName` and `SnapshotFormat` names more confusing that e.g. `ModuleName` and `DefaultSnapshotFormat`. Is there other code expecting these particular names, and if so can their doc comments reference that for explanation?

Reply:
It's cosmos code that expects this interface. Will try to add comments to all `ExtensionSnapshotter` interface methods.